README.md
# SixArm.com → Ruby → <br> Numeric #round, #floor, #ceil methods with precision

<!--header-open-->

[![Gem Version](https://badge.fury.io/rb/sixarm_ruby_numeric_round.svg)](http://badge.fury.io/rb/sixarm_ruby_numeric_round)
[![Build Status](https://travis-ci.org/SixArm/sixarm_ruby_numeric_round.png)](https://travis-ci.org/SixArm/sixarm_ruby_numeric_round)
[![Code Climate Maintainability](https://api.codeclimate.com/v1/badges/1cf9d84964d38caab79a/maintainability)](https://codeclimate.com/github/SixArm/sixarm_ruby_numeric_round/maintainability)

* Git: <https://github.com/sixarm/sixarm_ruby_numeric_round>
* Doc: <http://sixarm.com/sixarm_ruby_numeric_round/doc>
* Gem: <https://rubygems.org/gems/sixarm_ruby_numeric_round>
* Contact: Joel Parker Henderson, <joel@sixarm.com>
* Project: [changes](CHANGES.md), [license](LICENSE.md), [contributing](CONTRIBUTING.md).

<!--header-shut-->


## Introduction

Simple numeric rounding methods like #round, #floor, #ceil with various precisions.

For docs go to <http://sixarm.com/sixarm_ruby_numeric_round/doc>

Want to help? We're happy to get pull requests.


<!--install-open-->

## Install

### Gem

Run this command in your shell or terminal:

    gem install sixarm_ruby_numeric_round

Or add this to your Gemfile:

    gem 'sixarm_ruby_numeric_round'

### Require

To require the gem in your code:

    require 'sixarm_ruby_numeric_round'

<!--install-shut-->


## Examples

Round at a given decimal point:

    4.555.ceil_at(1)  #=> 4.6
    4.555.floor_at(1)  #=> 4.5
    4.555.round_at(1)  #=> 4.6
    4.555.truncate_at(1)  #=> 4.5

Round to a given precision:

    4.555.ceil_to(0.1)   #=> 4.6
    4.555.floor_to(0.1)  #=> 4.5
    4.555.round_to(0.1)  #=> 4.6
    4.555.truncate_to(0.1)  #=> 4.5

Conceptually, the methods for round, floor, ceil, and truncate will typically be for floating point numbers.
However, these methods can actually be for pretty much any Numeric object.
For example, one could round an Integer to the nearest kilo.


## Credit

These methods are based on the Facets library for Ruby and the developer named Trans. 

The #round_at and #round_to methods are exact copies and fully compatible with Facets. The rest of the methods are very similar code and fully co-existant with Facets. 

We're asking the Facets team to consider including the rest of the methods.


CHANGES.md
# Changes

* 2018-06-22 2.0.0 Update to Ruby 2.5, testing gems, etc.
* 2017-08-14 1.0.6 Update minitest expections
* 2017-08-13 1.0.5 Update to Ruby 2.4, spdx licenses, rake test, etc.
* 2015-07-18 1.0.4 Improve quality assurance
* 2015-07-07 1.0.3 Update gemspec to use file manifest
* 2012-09-01 1.0.2 Add #truncate_at and #truncate_to; improve floating division
* 2012-05-30 1.0.0 Create baesd on Facets methods by Trans


test/sixarm_ruby_numeric_round_test.rb
# -*- coding: utf-8 -*-
require "minitest/autorun"
require "simplecov"
SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter.new([
  SimpleCov::Formatter::HTMLFormatter,
])
SimpleCov.start
require "sixarm_ruby_numeric_round"


test/sixarm_ruby_numeric_round_test/float_test.rb
# -*- coding: utf-8 -*-
require "sixarm_ruby_numeric_round_test"

describe Float do

  describe "#ceil_at" do

    it "works" do
      4.555.ceil_at(0).must_equal 5.0
      4.555.ceil_at(1).must_equal 4.6
      4.555.ceil_at(2).must_equal 4.56
      4.555.ceil_at(3).must_equal 4.555
    end

  end

  describe "#ceil_to" do

    it "works" do
      4.555.ceil_to(1).must_equal 5.0
      4.555.ceil_to(0.1).must_equal 4.6
      4.555.ceil_to(0.01).must_equal 4.56
      4.555.ceil_to(0).must_equal 4.555
    end

  end

  describe "#floor_at" do

    it "works" do
      4.555.floor_at(0).must_equal 4.0
      4.555.floor_at(1).must_equal 4.5
      4.555.floor_at(2).must_equal 4.55
      4.555.floor_at(3).must_equal 4.555
    end

  end

  describe "#floor_to" do

    it "works" do
      4.555.floor_to(1).must_equal 4.0
      4.555.floor_to(0.1).must_equal 4.5
      4.555.floor_to(0.01).must_equal 4.55
      4.555.floor_to(0).must_equal 4.555
    end

  end

  describe "#round_at" do

    it "works" do
      4.555.round_at(0).must_equal 5.0
      4.555.round_at(1).must_equal 4.6
      4.555.round_at(2).must_equal 4.56
      4.555.round_at(3).must_equal 4.555
    end

  end

  describe "#round_to" do

    it "works" do
      4.555.round_to(1).must_equal 5
      4.555.round_to(0.1).must_equal 4.6
      4.555.round_to(0.01).must_equal 4.56
      4.555.round_to(0).must_equal 4.555
    end

  end

  describe "#truncate_at" do

    it "works" do
      4.555.truncate_at(0).must_equal 4.0
      4.555.truncate_at(1).must_equal 4.5
      4.555.truncate_at(2).must_equal 4.55
      4.555.truncate_at(3).must_equal 4.555
    end

  end

  describe "#truncate_to" do

    it "works" do
      4.555.truncate_to(1).must_equal 4.0
      4.555.truncate_to(0.1).must_equal 4.5
      4.555.truncate_to(0.01).must_equal 4.55
      4.555.truncate_to(0).must_equal 4.555
    end

  end

end









test/sixarm_ruby_numeric_round_test/numeric_test.rb
# -*- coding: utf-8 -*-
require "sixarm_ruby_numeric_round_test"

describe Numeric do

  describe "#ceil_at" do

    it "works" do
      4.555.ceil_at(0).must_equal 5.0
      4.555.ceil_at(1).must_equal 4.6
      4.555.ceil_at(2).must_equal 4.56
      4.555.ceil_at(3).must_equal 4.555
    end

  end

  describe "#ceil_to" do

    it "works" do
      4.555.ceil_to(1).must_equal 5.0
      4.555.ceil_to(0.1).must_equal 4.6
      4.555.ceil_to(0.01).must_equal 4.56
      4.555.ceil_to(0).must_equal 4.555
    end

  end

  describe "#floor_at" do

    it "works" do
      4.555.floor_at(0).must_equal 4.0
      4.555.floor_at(1).must_equal 4.5
      4.555.floor_at(2).must_equal 4.55
      4.555.floor_at(3).must_equal 4.555
    end

  end

  describe "#floor_to" do

    it "works" do
      4.555.floor_to(1).must_equal 4.0
      4.555.floor_to(0.1).must_equal 4.5
      4.555.floor_to(0.01).must_equal 4.55
      4.555.floor_to(0).must_equal 4.555
    end

  end

  describe "#round_at" do

    it "works" do
      4.555.round_at(0).must_equal 5.0
      4.555.round_at(1).must_equal 4.6
      4.555.round_at(2).must_equal 4.56
      4.555.round_at(3).must_equal 4.555
    end

  end

  describe "#round_to" do

    it "works" do
      4.555.round_to(1).must_equal 5
      4.555.round_to(0.1).must_equal 4.6
      4.555.round_to(0.01).must_equal 4.56
      4.555.round_to(0).must_equal 4.555
    end

  end

  describe "#truncate_at" do

    it "works" do
      4.555.truncate_at(0).must_equal 4.0
      4.555.truncate_at(1).must_equal 4.5
      4.555.truncate_at(2).must_equal 4.55
      4.555.truncate_at(3).must_equal 4.555
    end

  end

  describe "#truncate_to" do

    it "works" do
      4.555.truncate_to(1).must_equal 4.0
      4.555.truncate_to(0.1).must_equal 4.5
      4.555.truncate_to(0.01).must_equal 4.55
      4.555.truncate_to(0).must_equal 4.555
    end

  end

end









lib/sixarm_ruby_numeric_round/numeric.rb
# -*- coding: utf-8 -*-
=begin rdoc
Numeric extensions to calculate round numbers.
=end

class Numeric

  # Conceptually, ceil is expected to apply to floating point numbers.
  # However it can actually be applied to pretty much any Numeric object.
  # For example, one could ceil an Integer to the nearest kilo.
  #
  # See Float#ceil_at.

  def ceil_at(*args)
    to_f.ceil_at(*args)
  end

  # See Float#ceil_to.

  def ceil_to(*args)
    to_f.ceil_to(*args)
  end

  # Conceptually, floor is expected to apply to floating point numbers.
  # However it can actually be applied to pretty much any Numeric object.
  # For example, one could floor an Integer to the nearest kilo.
  #
  # See Float#floor_at.

  def floor_at(*args)
    to_f.floor_at(*args)
  end

  # See Float#floor_to.

  def floor_to(*args)
    to_f.floor_to(*args)
  end

  # Conceptually, round is expected to apply to floating point numbers.
  # However it can actually be applied to pretty much any Numeric object.
  # For example, one could round an Integer to the nearest kilo.
  #
  # See Float#round_at.

  def round_at(*args)
    to_f.round_at(*args)
  end

  # See Float#round_to.

  def round_to(*args)
    to_f.round_to(*args)
  end

  # Conceptually, truncate is expected to apply to floating point numbers.
  # However it can actually be applied to pretty much any Numeric object.
  # For example, one could truncate an Integer to the nearest kilo.
  #
  # See Float#truncate_at.

  def truncate_at(*args)
    to_f.truncate_at(*args)
  end

  # See Float#truncate_to.

  def truncate_to(*args)
    to_f.truncate_to(*args)
  end

end


lib/sixarm_ruby_numeric_round/float.rb
# -*- coding: utf-8 -*-
=begin rdoc
Float extensions to calculate round numbers.
=end

class Float

  # Ceil to the given decimal position.
  #
  #   4.555.ceil_at(0)  #=> 5.0
  #   4.555.ceil_at(1)  #=> 4.6
  #   4.555.ceil_at(2)  #=> 4.56
  #   4.555.ceil_at(3)  #=> 4.555
  #
  # CREDIT: Trans & Joel Parker Henderson

  def ceil_at( d ) #d=0
    (self * (10 ** d)).ceil.fdiv (10 ** d)
  end

  # Ceils to the nearest _n_th degree.
  #
  #   4.555.ceil_to(1)     #=> 5.0
  #   4.555.ceil_to(0.1)   #=> 4.6
  #   4.555.ceil_to(0.01)  #=> 4.56
  #   4.555.ceil_to(0)     #=> 4.555
  #
  # CREDIT: Trans & Joel Parker Henderson

  def ceil_to( n ) #n=1
    return self if n == 0
    (self * (1.0 / n)).ceil.fdiv (1.0 / n)
  end

  # Floor to the given decimal position.
  #
  #   4.555.floor_at(0)  #=> 4.0
  #   4.555.floor_at(1)  #=> 4.5
  #   4.555.floor_at(2)  #=> 4.55
  #   4.555.floor_at(3)  #=> 4.555
  #
  # CREDIT: Trans & Joel Parker Henderson

  def floor_at( d ) #d=0
    (self * (10 ** d)).floor.fdiv (10 ** d)
  end

  # Floors to the nearest _n_th degree.
  #
  #   4.555.floor_to(1)     #=> 4.0
  #   4.555.floor_to(0.1)   #=> 4.5
  #   4.555.floor_to(0.01)  #=> 4.55
  #   4.555.floor_to(0)     #=> 4.555
  #
  # CREDIT: Trans & Joel Parker Henderson

  def floor_to( n ) #n=1
    return self if n == 0
    (self * (1.0 / n)).floor.fdiv (1.0 / n)
  end

  # Round to the given decimal position.
  #
  #   4.555.round_at(0)  #=> 5.0
  #   4.555.round_at(1)  #=> 4.6
  #   4.555.round_at(2)  #=> 4.56
  #   4.555.round_at(3)  #=> 4.555
  #
  # CREDIT: Trans & Joel Parker Henderson

  def round_at( d ) #d=0
    (self * (10 ** d)).round.fdiv (10 ** d)
  end

  # Rounds to the nearest _n_th degree.
  #
  #   4.555.round_to(1)     #=> 5
  #   4.555.round_to(0.1)   #=> 4.6
  #   4.555.round_to(0.01)  #=> 4.56
  #   4.555.round_to(0)     #=> 4.555
  #
  # CREDIT: Trans & Joel Parker Henderson

  def round_to( n ) #n=1
    return self if n == 0
    (self * (1.0 / n)).round.fdiv (1.0 / n)
  end

  # Truncate to the given decimal position.
  #
  #   4.555.truncate_at(0)  #=> 4.0
  #   4.555.truncate_at(1)  #=> 4.5
  #   4.555.truncate_at(2)  #=> 4.55
  #   4.555.truncate_at(3)  #=> 4.555
  #
  # CREDIT: Trans & Joel Parker Henderson

  def truncate_at( d ) #d=0
    (self * (10 ** d)).truncate.fdiv (10 ** d)
  end

  # Truncates to the nearest _n_th degree.
  #
  #   4.555.truncate_to(1)     #=> 4.0
  #   4.555.truncate_to(0.1)   #=> 4.5
  #   4.555.truncate_to(0.01)  #=> 4.55
  #   4.555.truncate_to(0)     #=> 4.555
  #
  # CREDIT: Trans & Joel Parker Henderson

  def truncate_to( n ) #n=1
    return self if n == 0
    (self * (1.0 / n)).truncate.fdiv (1.0 / n)
  end

end
